rfc-capi-for-kvm-capk/README.md

```diff
@@ -0,0 +1,425 @@
+
+Current features of `kvm-operator` (with a ✅ if we decided to keep it in CAPI):
+- [ ] configure the registry: we can set an auth token, domain and mirrors (we use that for quay for example I think); make the daemon run behind a proxy
+- [ ] configure workload cluster public key for ssh
+- [x] configure calico: CALICO_IPV4POOL_CIDR env var for calico container, CNI_MT env var for calico init container, clusterCIDR for kube-proxy)
+- [x] configure DOCKER_OPT_BIP (--bip flag) for docker daemon
+- [ ] configure domain and prefix for etcd
+- [ ] apiserver flags
+	- [ ] - configure oidc
+	- [ ] --etcd-prefix
+	- [ ] --service-cluster-ip-range
+	- [ ] --secure-port
+	- [ ] --requestheader-allowed-names
+- [ ] kubelet configuration
+	- [ ] clusterDNS
+	- [ ] clusterDomain
+	- [ ] --node-labels (flag)
+- [ ] kube-proxy
+	- [ ] conntrack.maxPerCore
+- [ ] ignition
+	- [x] dns and ntp servers
+	- [x] ssh user list
+- [ ] iscsi initiator name (masters)
+- [ ] configure storage type (host path/persistent)
+- [x] master and worker node pods: cpu and memory limits,
+- [x] size of kubelet and docker volume
+- [x] automatically delete TC node pods when NotReady for too long
+- [x] dead endpoints: don't route traffic to non-ready nodes
+- [x] vertical pod autoscaler for operator pod
+
+#### Features to drop
+- Done by the bootstrap provider in future: 
+    - kube-proxy configuration
+    - kubelet configuration
+    - api-server flags
+    - etcd configuration
+    - ssh server public keys
+    - registry configuration and prefix
+- Calico configuration
+    - Handled by cluster-apps-operator
+- Node IP range: Will be defined by the Manangment cluster POD IP range as we will use kubernetes native networking.
+
+
+### Implementation
+
+#### DNS
+
+The WC Kubernetes API is reachable externally (i.e. for GS engineers or customers) via an ingress in the MC of the form `https://<cluster id>.k8s.<installation base domain>`. This DNS name should resolve to the MC ingress IP.
+
+Inside the WC, control plane nodes should connect to etcd and k8s API via localhost. Workers should use a control plane endpoint of `https://<cluster id>.k8s.<installation base domain>` or `control-plane.<cluster id>.svc` depending on whether external DNS or MC CoreDNS is used for WC node DNS resolution (respectively). This will be determined during implementation.
+
+#### Machine OS Images
+
+The CAPI-native method for OS images is using an image builder to create a different image for each Kubernetes version. This implies a lot of complexity and is a big departure from our current approach of configuring nodes using ignition. During development we will use kubeadm pre- and post- commands to provision nodes and then decide if we want to go through the process of setting up an image building pipeline.
```

Comment:
Images built by image builder should be available soon (for flatcar as well) - so I don't think this would be a massive change for us? Maybe I am wrong.

Reply:
That would be great if we don't have to maintain it ourselves. Any docs about that?

Reply:
There are some upstream docs but there has been internal discussion with kinvolk:
https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/README-flatcar.md

We would need to check how much effort into image building we might need to put in. But upstream does have open PRs to add ignition support:
https://github.com/kubernetes-sigs/cluster-api/pull/4172

Reply:
Thanks. I added links to these in the RFC and mentioned that we would try to stay on top of what's going on there.